subsys/pm/pm.c

```diff
@@ -149,10 +175,13 @@ void pm_system_resume(void)
 	 * and it may schedule another thread.
 	 */
 	if (atomic_test_and_clear_bit(z_post_ops_required, id)) {
+		/* Exiting low-power state */
+		pm_state_notify(PM_STATE_EXIT);
 		pm_state_exit_post_ops(z_cpus_pm_state[id].state, z_cpus_pm_state[id].substate_id);
-		pm_state_notify(false);
```

Comment:
Note to self:

On STM32 we need to reinitialize peripherals after we've reinitialized clocks, but before we re-enable interrupts.
Both of these occur inside of `pm_state_exit_post_ops`, thus we should probably rethink the approach here.

Reply:
Isn't the case of re-initialize clocks in `pm_state_set` just before leaving this function. e.g:
```

/* Invoke Low Power/System Off specific Tasks */
void pm_state_set(enum pm_state state, uint8_t substate_id)
{
	if (state != PM_STATE_SUSPEND_TO_IDLE) {
		LOG_DBG("Unsupported power state %u", state);
		return;
	}

	/* ensure the proper wake-up system clock */
	LL_RCC_SetClkAfterWakeFromStop(RCC_STOP_WAKEUPCLOCK_SELECTED);

	switch (substate_id) {
	case 1: /* this corresponds to the STOP0 mode: */
		/* enter STOP0 mode */
		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP0);
		break;
	case 2: /* this corresponds to the STOP1 mode: */
		/* enter STOP1 mode */
		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP1);
		break;
	case 3: /* this corresponds to the STOP2 mode: */
#ifdef PWR_CR1_RRSTP
		LL_PWR_DisableSRAM3Retention();
#endif /* PWR_CR1_RRSTP */
		/* enter STOP2 mode */
		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP2);
		break;
	default:
		LOG_DBG("Unsupported power state substate-id %u",
			substate_id);
		break;
	}

	LL_LPM_EnableDeepSleep();
	/* enter SLEEP mode : WFE or WFI */
	k_cpu_idle();

       /* We have just waken up from idle, re-initialize clock */
	stm32_clock_control_init(NULL);
}
```

Then you have notification callback after the clock being initialized and before restoring interruptions.

Reply:
Moving clock reinit to after idle in `pm_state_set` on STM32 won't have the intended affect, as the `pm_system_resume()` logic (called via idle's `z_pm_save_idle_exit()`) is run in the ISR of the wakeup source, and not in the idle thread.

I was thinking of refactoring the IRQ unlocking: it's pretty unnecessary to do this in the `pm_state_exit_post_ops()`, as *every* platform needs to reenable IRQs after wakeup to reverse the `irq_lock()` done by idle.
On several platforms the post_ops are used exclusively for reenabling IRQs, so maybe extending the API to could deal with it on the architecture-level instead of for each SoC-series could be a good way to remove this redundancy. It'd also allow us to wedge the notifier callback in between the post_ops and irq_unlock.

Reply:
For now I've moved the notifier to after post_ops. The IRQ changes are larger as there are discrepancies in some post_ops logic that block it from being implemented cleanly. I'll create a draft PR for it when ready.